Improve and unify connection configuration process

ivan-podorozhnyi opened the pull request:
`config_draft.properties` is supposed to store data about remote virtual device.
reader.py module contains SshUserParser class that allows to read, parse data from configuration file and create a user based on data from configuration file.

Affects #5

AntonKarabaza commented:
@extsoft There are some doubts. May be it's better to create one method `get_property(property: str)` instead of methods for each property. What do you think?

extsoft commented:
> @extsoft There are some doubts. May be it's better to create one method get_property(property: str) instead of methods for each property. What do you think?

@AntonKarabaza  It's a terrible idea to have `property_reader.py` at all.

@AntonKarabaza @ivan-podorozhnyi which object you would like to get from properties file? And were it has to be used? I don't see any usage right now.

ivan-podorozhnyi commented:
@extsoft PropertyReader instance will be used to pass configuration to setup fixtures in `conftest.py`. I will implement these changes after this pull request approval.
I think it's a good idea to use configuration file because hardcoded settings in tests look awful.

extsoft commented:
@ivan-podorozhnyi 
> which object you would like to get from properties file?

ivan-podorozhnyi commented:
@extsoft I see what you mean. I will add another class called `Property` and update whole pull request.

extsoft commented:
I’m not sure that you understand what I mean. Why we need Property class? And you didn’t answer my previous question.
> On Jul 11, 2017, at 11:46, Ivan <notifications@github.com> wrote:
> 
> @extsoft <https://github.com/extsoft> I see what you mean. I will add another class called Property and update whole pull request.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/ivan-podorozhnyi/ssh_tasks/pull/6#issuecomment-314376570>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFiaI0R8JEvJf58_2NFzo1hK-FvBd50Tks5sMzZOgaJpZM4OQ-hk>.
>

ivan-podorozhnyi commented:
@extsoft According to your recommendations I made some changes:
1) ConfiguredUser class was implemented to create user based on data from configuration file. 
2) ConfigReader class was created to supply data from configuration file.
Please let me know if it's ok.

extsoft commented on ssh_automated_tests/core/user.py:
```diff
@@ -16,3 +20,14 @@ def make_connection_data(self):
+class ConfiguredUser(User):
+    def __init__(self, reader: ConfigReader):
+        self._reader = reader
+        self._user = SshUser(self._reader.read('user_name'),
```
There is a work in the constructor. That's a bad idea.

extsoft commented on ssh_automated_tests/core/property_reader.py:
```diff
@@ -0,0 +1,18 @@
+        self._config = configparser.ConfigParser()
+        self._source_file = file
+
+    def read(self, parameter: str):
```
It sounds a bit crazy that the `read` method returns something. May we can identify a better name for this method?

extsoft commented on config_draft.properties:
```diff
@@ -0,0 +1,7 @@
+[connection]
+remote_device = your device address here
+#for example 127.0.0.1
```
It's great to have comments or explanations before the property definition.

extsoft commented on config_draft.properties:
```diff
@@ -0,0 +1,7 @@
+user_name = ivan
+device_port = 3022
+key = ../id_rsa
+# relative path to your ssh key
```
What is a starting point to define relative place?

extsoft commented on config_draft.properties:
```diff
@@ -0,0 +1,7 @@
+[connection]
+remote_device = your device address here
```
May let's rename to `ip-address` or something like that. It has to be self-explained.

ivan-podorozhnyi commented:
@extsoft I made some changes according to your recommendations:
1) renamed property names in configuration file and changed comments for it;
2) reworked the principle of parsing data from config file (aded SshUserParser class).
Let me know your opinion.

ivan-podorozhnyi commented:
@extsoft The idea of using configuration file was implemented in existing tests.

extsoft commented on ssh_automated_tests/core/user.py:
```diff
@@ -16,3 +20,13 @@ def make_connection_data(self):
                 "username": self._name, "key": self._key}
 
+
+class SshUserParser(object):
```
@ivan-podorozhnyi Where do you use this class?

ivan-podorozhnyi replied:
@extsoft Well, seems I got confused around this implementation.
I see that it's possible to avoid creating extra classes like `SshUserParser`.
What do you think? Should I remove that class or use it instead?

extsoft replied:
Please propose me at least 3 solutions. Then we can discuss them.

ivan-podorozhnyi replied:
@extsoft Ok, I have several ideas:
1) Avoid creating extra classes and create user instance using `ConfgReader` class.

```
    cfg_reader = ConfigReader(RawFile('config_draft.properties'))

    user = SshUser(name=cfg_reader.get_value('connection', 'user_name'),
                   port=int(cfg_reader.get_value('connection', 'port')),
                   key=cfg_reader.get_value('connection', 'key'),
                   address=cfg_reader.get_value('connection', 'ip_address'))
```
2) Use `SshUserParser` to create user and hide some work inside it`s method.

```
    parser = SshUserParser(ConfigReader(RawFile('config_draft.properties')))

    user = parser.parse_user('connection', 'user_name', 'port', 'key', 'ip_address')
```

3) Rework previous way to hide all parameters inside class method.

```
parser = SshUserParser(ConfigReader(RawFile('config_draft.properties')))

user = parser.parse_user()
```


4)  Change the principle of parsing inside `ConfigReader` class to use built in `open()` and `*.readline()` functions and for each loops to parse configuration file. This can avoid adding `[connection]` section to configuration file. Code in `conftest.py` will look similar to previous.
I think it will be possible to use other file extensions for configurations (for example json, ini, yaml, xml or txt).

```
    parser = SshUserParser(ConfigReader(RawFile('config_draft.txt')))

    user = parser.parse_user()
```

ivan-podorozhnyi commented:
@extsoft I changed the structure according to our morning conversation.
In final version SshUserFromPropFile and PropertiesFile classes are used to read properties file and prepare all data required for connection.
Please let me know your opinion.

extsoft commented:
@ivan-podorozhnyi great job!
